File: backend/app/services/rag.py
"""Grounded retrieval, citation construction, and answer generation."""

from __future__ import annotations

from dataclasses import dataclass
from uuid import UUID

from app.constants.prompts import RAG_SYSTEM_PROMPT, RAG_USER_PROMPT
from app.core.config import Settings
from app.core.logging import get_logger, log_process_finished, log_process_started
from app.integrations.embeddings import EmbeddingService
from app.integrations.ollama import OllamaChatService
from app.integrations.qdrant import QdrantVectorStore
from app.schemas.documents import Citation

logger = get_logger(__name__)


@dataclass(frozen=True, slots=True)
class RetrievalResult:
    """Formatted context and its ordered citations."""

    context: str
    citations: list[Citation]


class RAGService:
    """End-to-end retrieval with tenant and optional document filters."""

    def __init__(
        self,
        settings: Settings,
        embeddings: EmbeddingService,
        vector_store: QdrantVectorStore,
        llm: OllamaChatService,
    ) -> None:
        self.settings = settings
        self.embeddings = embeddings
        self.vector_store = vector_store
        self.llm = llm

    async def retrieve(
        self,
        *,
        user_id: UUID,
        query: str,
        document_ids: list[UUID] | None,
        top_k: int | None = None,
    ) -> RetrievalResult:
        """Embed a question and return grounded source chunks."""
        log_process_started(
            logger,
            "Retrieve relevant chunks",
            user_id=str(user_id),
            selected_documents=len(document_ids or []),
            top_k=top_k or self.settings.rag_top_k,
        )
        logger.info(
            "rag_retrieval_started — embedding query and searching Qdrant",
            user_id=str(user_id),
            query_preview=query[:120],
            document_ids=[str(d) for d in document_ids] if document_ids else "all",
            top_k=top_k or self.settings.rag_top_k,
        )
        query_vector = await self.embeddings.embed_query(query)
        hits = await self.vector_store.search(
            user_id=user_id,
            query_vector=query_vector,
            document_ids=document_ids,
            top_k=top_k or self.settings.rag_top_k,
            score_threshold=self.settings.rag_score_threshold,
        )
        log_process_started(
            logger,
            "Build grounded context",
            user_id=str(user_id),
            hits=len(hits),
        )
        citations: list[Citation] = []
        context_blocks: list[str] = []
        for index, hit in enumerate(hits, start=1):
            payload = hit.payload
            reference = f"C{index}"
            text = str(payload.get("chunk_text", ""))
            citation = Citation(
                reference=reference,
                document_id=UUID(str(payload["document_id"])),
                document_name=str(payload.get("filename", "Unknown document")),
                page_number=(
                    int(payload["page_number"]) if payload.get("page_number") is not None else None
                ),
                chunk_id=UUID(str(payload["chunk_id"])),
                chunk_index=int(payload.get("chunk_index", 0)),
                score=hit.score,
                excerpt=text[:500],
            )
            citations.append(citation)
            page = citation.page_number if citation.page_number is not None else "N/A"
            context_blocks.append(
                f"[{reference}] Document: {citation.document_name}; "
                f"Page: {page}; Chunk: {citation.chunk_index}\n{text}"
            )
        logger.info(
            "rag_retrieval_completed — context built, ready for generation",
            user_id=str(user_id),
            hits=len(hits),
            citations=len(citations),
            context_chars=sum(len(b) for b in context_blocks),
        )
        context = "\n\n".join(context_blocks)
        log_process_finished(
            logger,
            "Build grounded context",
            user_id=str(user_id),
            citations=len(citations),
            context_chars=len(context),
        )
        log_process_finished(
            logger,
            "Retrieve relevant chunks",
            user_id=str(user_id),
            hits=len(hits),
            citations=len(citations),
        )
        return RetrievalResult(context=context, citations=citations)

    async def answer(self, *, question: str, retrieval: RetrievalResult) -> str:
        """Generate an answer constrained by retrieved context."""
        log_process_started(
            logger,
            "Generate grounded answer",
            citations=len(retrieval.citations),
        )
        if not retrieval.citations:
            logger.info("rag_answer_skipped — no citations found, returning fallback message")
            answer = "I could not find enough information in the selected documents."
            log_process_finished(
                logger,
                "Generate grounded answer",
                citations=0,
                answer_chars=len(answer),
            )
            return answer
        logger.info(
            "rag_answer_started — sending context to LLM for grounded answer",
            citations=len(retrieval.citations),
        )
        result = await self.llm.generate(
            RAG_SYSTEM_PROMPT,
            RAG_USER_PROMPT.format(question=question, context=retrieval.context),
        )
        logger.info("rag_answer_done", answer_chars=len(result))
        log_process_finished(
            logger,
            "Generate grounded answer",
            citations=len(retrieval.citations),
            answer_chars=len(result),
        )
        return result

File: backend/src/services/rag.js
import { config } from "../config.js";
import { RAG_SYSTEM_PROMPT, RAG_USER_PROMPT } from "../constants/prompts.js";
import { embeddings, llm } from "../integrations/ollama.js";
import { vectorStore } from "../integrations/qdrant.js";
import { log, logProcessFinished, logProcessStarted } from "../logger.js";

export class RAGService {
  async retrieve({ userId, query, documentIds, topK = config.ragTopK }) {
    logProcessStarted("Retrieve RAG context", {
      user_id: userId,
      query,
      document_ids: documentIds ?? null,
      top_k: topK,
      score_threshold: config.ragScoreThreshold,
    });
    const queryVector = await embeddings.embedQuery(query);
    log("info", "rag_query_embedding_created", {
      message: "RAG query embedding created.",
      user_id: userId,
      dimensions: queryVector.length,
    });
    const hits = await vectorStore.search({
      userId,
      queryVector,
      documentIds,
      topK,
      scoreThreshold: config.ragScoreThreshold,
    });
    log("info", "rag_vector_hits_received", {
      message: "RAG vector search hits received.",
      user_id: userId,
      hits: hits.length,
      scores: hits.map((hit) => hit.score),
    });
    const citations = [];
    const contextBlocks = [];
    for (const [index, hit] of hits.entries()) {
      const reference = `C${index + 1}`;
      const text = String(hit.payload.chunk_text ?? "");
      const citation = {
        reference,
        document_id: String(hit.payload.document_id),
        document_name: String(hit.payload.filename ?? "Unknown document"),
        page_number:
          hit.payload.page_number === null || hit.payload.page_number === undefined
            ? null
            : Number(hit.payload.page_number),
        chunk_id: String(hit.payload.chunk_id),
        chunk_index: Number(hit.payload.chunk_index ?? 0),
        score: Number(hit.score),
        excerpt: text.slice(0, 500),
      };
      citations.push(citation);
      contextBlocks.push(
        `[${reference}] Document: ${citation.document_name}; Page: ${
          citation.page_number ?? "N/A"
        }; Chunk: ${citation.chunk_index}\n${text}`,
      );
    }
    logProcessFinished("Retrieve RAG context", {
      user_id: userId,
      citations: citations.length,
      context_chars: contextBlocks.join("\n\n").length,
    });
    return { context: contextBlocks.join("\n\n"), citations };
  }

  async answer({ question, retrieval }) {
    logProcessStarted("Generate RAG answer", {
      question,
      citations: retrieval.citations.length,
      context_chars: retrieval.context.length,
    });
    if (retrieval.citations.length === 0) {
      const fallback = "I could not find enough information in the selected documents.";
      logProcessFinished("Generate RAG answer", { answer: fallback });
      return fallback;
    }
    const answer = await llm.generate(
      RAG_SYSTEM_PROMPT,
      RAG_USER_PROMPT({ question, context: retrieval.context }),
    );
    logProcessFinished("Generate RAG answer", { answer });
    return answer;
  }
}

export const ragService = new RAGService();
